source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -72,90 +175,127 @@ void Filter::onReadWorker() {
     if (line_parts.size() != 6) {
       throw EnvoyException("failed to read proxy protocol");
     }
-
-    Network::Address::IpVersion protocol_version;
-    Network::Address::InstanceConstSharedPtr remote_address;
-    Network::Address::InstanceConstSharedPtr local_address;
+    hdr.valid_ = true;
 
     // TODO(gsagula): parseInternetAddressAndPort() could be modified to take two string_view
     // arguments, so we can eliminate allocation here.
     if (line_parts[1] == "TCP4") {
-      protocol_version = Network::Address::IpVersion::v4;
-      remote_address = Network::Utility::parseInternetAddressAndPort(
+      hdr.protocol_version = Network::Address::IpVersion::v4;
+      hdr.remote_address = Network::Utility::parseInternetAddressAndPort(
           std::string{line_parts[2]} + ":" + std::string{line_parts[4]});
-      local_address = Network::Utility::parseInternetAddressAndPort(
+      hdr.local_address = Network::Utility::parseInternetAddressAndPort(
           std::string{line_parts[3]} + ":" + std::string{line_parts[5]});
     } else if (line_parts[1] == "TCP6") {
-      protocol_version = Network::Address::IpVersion::v6;
-      remote_address = Network::Utility::parseInternetAddressAndPort(
+      hdr.protocol_version = Network::Address::IpVersion::v6;
+      hdr.remote_address = Network::Utility::parseInternetAddressAndPort(
           "[" + std::string{line_parts[2]} + "]:" + std::string{line_parts[4]});
-      local_address = Network::Utility::parseInternetAddressAndPort(
+      hdr.local_address = Network::Utility::parseInternetAddressAndPort(
           "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]});
     } else {
       throw EnvoyException("failed to read proxy protocol");
     }
+  }
+}
 
-    // Error check the source and destination fields. Most errors are caught by the address
-    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
-    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
-    // address.
-    const auto remote_version = remote_address->ip()->version();
-    const auto local_version = local_address->ip()->version();
-    if (remote_version != protocol_version || local_version != protocol_version) {
-      throw EnvoyException("failed to read proxy protocol");
-    }
-    // Check that both addresses are valid unicast addresses, as required for TCP
-    if (!remote_address->ip()->isUnicastAddress() || !local_address->ip()->isUnicastAddress()) {
-      throw EnvoyException("failed to read proxy protocol");
+bool Filter::readProxyHeader(int fd, WireHeader& hdr) {
+  while (buf_off_ < MAX_PROXY_PROTO_LEN) {
+    int bytes_avail;
+
+    if (ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
```

Comment:
Can this ever reasonably fail? Should this be a RELEASE_ASSERT?

Reply:
So i'm not sure what RELEASE_ASSERT is for.
The only case this can fail is the race condition where the socket is closed (RST from far side) between waking up and reaching here.

the same is true for the recv MSG_PEEK. the kernel could tear down the PCB (so_linger?), but otherwise the read will succeed.

it seems RELEASE_ASSERT is one that is always on (not NDEBUG conditional) and causes an abort(). I think we don't want that here, we want to unwind and close if the socket is removed during the listener phase.